### PR TITLE
Move compile time settings from makefile to lmpconfig.h header

### PR DESCRIPTION
**Pull request by akohlmey**
## Purpose

Several compile time settings, most prominently -DLAMMPS_BIGBIG/-DLAMMPS_SMALLBIG/-DLAMMPS_SMALLSMALL have to be kept consistent across different parts (e.g. in src and lib folder) of LAMMPS and particularly when installing the LAMMPS library interface from a pre-compiled package.

To integrate LAMMPS more cleanly into the typical Linux/Unix-like file system structure when using a package manager software, the library should be installed in /usr/lib64/ and headers in /usr/include/lammps. For the C-style library interface, only library.h and lmpconfig.h are required and
only library.h would need to be included via `#include <lammps/library.h>`.
if also C++ should be supported, additional header files should be installed, e.g. lammps.h and several constituent base classes in lammps.h (e.g. input.h, atom.h etc.) and beyond.

lmpconfig.h is not part of the distribution and will be automatically created with default setting by the makefile in src, via calling `sh Config.sh default`. This can be customized via `make config`.

## Author(s)

Axel Kohlmeyer (Temple U)

## Backward Compatibility

The auto-generated header file will error out, if the same property is also set in a makefile and the user instructed to remove that define from the makefile and use `make config` instead.

## Implementation Notes

N/A

## Post Submission Checklist

_Please check the fields below as they are completed_
- [ ] The feature or features in this pull request is complete
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] One or more example input decks are included
- [ ] The source code follows the LAMMPS formatting guidelines

### CODE REVIEW

File: examples/COUPLE/simple/simple.cpp [hunk 1]
```diff
@@ -19,15 +19,16 @@
...
-#include "input.h"
-#include "atom.h"
-#include "library.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <mpi.h>
+
+// these are LAMMPS include files
+#include <lammps/lammps.h>
```
Comment (junghans) [inline examples/COUPLE/simple/simple.cpp]: @akohlmey: And this example is actually the reason why the code (see https://github.com/lammps/lammps/pull/627#issuecomment-325409215) used `lmp = new LAMMPS()`.

**Comment by junghans** [comment]
Hmm, if `library.h` includes `lmpconfig.h` then is needs to generated and installed too (by CMake).

File: src/Config.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            1) intsize='LAMMPS_SMALLSMALL' ;;
+            2|"") intsize='LAMMPS_SMALLBIG'   ;;
+            3) intsize='LAMMPS_BIGBIG'     ;; 
+            *) echo "Invalid input: ${choice}"     ;;
+        esac
+    done
+    echo "using ${intsize}"
+fi
+
+echo "#define ${intsize}" >> ${tmpfile}
```
Comment (junghans) [inline src/Config.sh]: This might be a good place for `LAMMPS_EXCEPTIONS`, too.

**Comment by junghans** [comment]
Maybe we could generate the `lmpconfig.h`from an autotools-style templated `lmpconfig.h.in`:
```
...
#define @INTSIZE@
...
```
instead of a shell script, so that CMake and make can use the same file to start from.

In Make one would do:
```make
always: ;

.PHONY: always

#only update .define is flags change
.defines: always
	@[ -f $@ ] || touch $@
	@echo "DEFINES=$(LMP_INC)" | cmp -s $@ - || echo "DEFINES=$(LMP_INC)" > $@

lmpconfig.h: lmpconfig.h.in .defines
	sed -e 's/@INTSIZE@/$(INTSIZE)/'  -e ... $< > $@
```
and in CMake:
```cmake
configure_file(lmpconfig.h.in lmpconfig.h @ONLY)
```

**Comment by akohlmey** [comment]
@junghans the `Config.sh` shell script will remain. This allows a much more robust handling of the situation and can be written to guarantee that a functional lmpconfig.h is written (or none).
However, processing a template file is an option i have pondered, but not yet looked into for simplicity.
If processing a template file from CMake is desirable, then i will look into it. The is no obvious reason why the Config.sh script cannot process such a template instead of generating the file completely.

**Comment by junghans** [comment]
@akohlmey a template would help a lot to keep thing in sync easier.

**Comment by akohlmey** [comment]
@junghans how would CMake handle the case?
```
#define LAMMPS_GZIP
```
vs.
```
/* #undef LAMMPS_GZIP */
```

**Comment by junghans** [comment]
it cannot handle that: https://cmake.org/cmake/help/v3.0/command/configure_file.html
There are two options:
```
#define BLA @FOO@
```
or
```
#cmakedefine BLA
```
where the latter is a bit too CMake specific for us.

**Comment by junghans** [comment]
I could make that work!

**Comment by akohlmey** [comment]
@junghans great! i've rebased the branch on the latest master and did a forced push, so that all merged CMake updates are included. i will complete the settings that i can add in the make build environment (anything that requires changes to the linker flags may be added later, but not now) and convert it to use a template include file. we can then synchronize this here with CMake.

**Comment by junghans** [comment]
Huh? Where is it?

**Comment by akohlmey** [comment]
@junghans i wrote "i will" not "i have".

**Comment by akohlmey** [comment]
will be revisiting this topic at a later point, after we have some practical experience with CMake use "in the wild" following the upcoming stable release. by that time things will have changed so much, that it will be cleaner to start over. thus closing this PR for now.

**Pull request closed without merging**